akka-stream-tests/src/test/scala/akka/stream/scaladsl/RetryFlowSpec.scala

```diff
@@ -0,0 +1,572 @@
+          .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.right)
+      val (killSwitch, sink) = TestSource
+        .probe[(Int, Int)]
+        .viaMat(RetryFlow.withBackoffAndContext(10.millis, 5.second, 0, 3, innerFlow)(incrementFailedValues))(
+          Keep.right)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      killSwitch.abort(FailedElem.failed.get)
+
+      sink.request(1)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    val stuckForeverRetrying =
+      RetryFlow.withBackoffAndContext(10.millis, 5.seconds, 0d, 3, failAllValuesFlow)(alwaysRecoveringFunc)
+
+    "tolerate killswitch abort before the RetryFlow while on retry spin" in {
+      val ((source, killSwitch), sink) = TestSource
+        .probe[Int]
+        .via(Utils.delayCancellation(10.seconds))
+        .viaMat(KillSwitches.single[Int])(Keep.both)
+        .map(i => (i, i))
+        .via(stuckForeverRetrying)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1)
+      sink.expectNoMessage()
+
+      killSwitch.abort(FailedElem.failed.get)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    "tolerate killswitch abort on the inner flow while on retry spin" in {
+      val ((source, killSwitch), sink) = TestSource
+        .probe[Int]
+        .map(i => (i, i))
+        .viaMat(
+          RetryFlow.withBackoffAndContext(
+            10.millis,
+            5.seconds,
+            0d,
+            3,
+            failAllValuesFlow
+              .via(Utils.delayCancellation(10.seconds))
+              .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.right))(alwaysRecoveringFunc))(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      source.sendNext(1)
+      sink.expectNoMessage()
+      killSwitch.abort(FailedElem.failed.get)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    "tolerate killswitch abort after the RetryFlow while on retry spin" in {
+      val ((source, killSwitch), sink) = TestSource
+        .probe[Int]
+        .map(i => (i, i))
+        .via(stuckForeverRetrying)
+        .via(Utils.delayCancellation(10.seconds))
+        .viaMat(KillSwitches.single[(Try[Int], Int)])(Keep.both)
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1)
+      sink.expectNoMessage()
+
+      killSwitch.abort(FailedElem.failed.get)
+      sink.expectError(FailedElem.failed.get)
+    }
+
+    "finish only after processing all elements in stream" in {
+      val (source, sink) = TestSource
+        .probe[(Int, Int)]
+        .via(RetryFlow.withBackoffAndContext(10.millis, 5.seconds, 0d, 3, failEvenValuesFlow)(incrementFailedValues))
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+
+      source.sendNext(1 -> 0)
+      source.sendNext(3 -> 0)
+      source.sendNext(2 -> 0)
+      source.sendComplete()
```

Comment:
Since the operator itself doesn't do any buffering this will just put those in a buffer elsewhere, so just testing with one element and then completing would be just as good.